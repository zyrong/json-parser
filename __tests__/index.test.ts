import { resolve } from 'path';
import { readdirSync } from 'fs'
import fs from 'fs/promises'
import jsonParse from '../src/index'

const FIXTURES_DIR = resolve(__dirname, './fixtures')
const TEST_TIMEOUT = 10000;

describe('fixtures', () => {
  const filenames = readdirSync(FIXTURES_DIR)

  it.each(filenames)(
    'test %s',
    async (filename) => {
      const jsonPath = resolve(FIXTURES_DIR, filename);
      const jsonBuf = await fs.readFile(jsonPath)
      const json = jsonBuf.toString()

      const visitor = jsonParse(json)
      if (visitor) {
        expect(visitor.body).toMatchSnapshot()
      } else {
        expect(visitor).toBe(null)
      }
    },
    TEST_TIMEOUT,
  );
})