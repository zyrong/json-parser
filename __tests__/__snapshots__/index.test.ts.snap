// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures test all.json 1`] = `
Object {
  "key": null,
  "parent": null,
  "properties": Array [
    Object {
      "key": Object {
        "range": Object {
          "end": 16,
          "start": 5,
        },
        "value": "empty_string",
      },
      "parent": [Circular],
      "type": "string",
      "value": Object {
        "code": "",
        "range": Object {
          "end": 20,
          "start": 21,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 32,
          "start": 27,
        },
        "value": "string",
      },
      "parent": [Circular],
      "type": "string",
      "value": Object {
        "code": "string",
        "range": Object {
          "end": 42,
          "start": 37,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 50,
          "start": 49,
        },
        "value": "\\\\\\"",
      },
      "parent": [Circular],
      "type": "string",
      "value": Object {
        "code": "\\\\\\"",
        "range": Object {
          "end": 55,
          "start": 54,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 67,
          "start": 62,
        },
        "value": "number",
      },
      "parent": [Circular],
      "type": "number",
      "value": Object {
        "code": "0",
        "range": Object {
          "end": 71,
          "start": 71,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 80,
          "start": 77,
        },
        "value": "null",
      },
      "parent": [Circular],
      "type": "null",
      "value": Object {
        "code": "null",
        "range": Object {
          "end": 87,
          "start": 84,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 97,
          "start": 93,
        },
        "value": "false",
      },
      "parent": [Circular],
      "type": "boolean",
      "value": Object {
        "code": "false",
        "range": Object {
          "end": 104,
          "start": 100,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 113,
          "start": 110,
        },
        "value": "true",
      },
      "parent": [Circular],
      "type": "boolean",
      "value": Object {
        "code": "true",
        "range": Object {
          "end": 120,
          "start": 117,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 130,
          "start": 126,
        },
        "value": "array",
      },
      "parent": [Circular],
      "properties": Array [],
      "type": "array",
      "value": Object {
        "code": "[]",
        "range": Object {
          "end": 135,
          "start": 134,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 146,
          "start": 141,
        },
        "value": "object",
      },
      "parent": [Circular],
      "properties": Array [],
      "type": "object",
      "value": Object {
        "code": "{}",
        "range": Object {
          "end": 151,
          "start": 150,
        },
      },
    },
    Object {
      "key": Object {
        "range": Object {
          "end": 160,
          "start": 157,
        },
        "value": "deep",
      },
      "parent": [Circular],
      "properties": Array [
        Object {
          "key": 0,
          "parent": [Circular],
          "properties": Array [
            Object {
              "key": Object {
                "range": Object {
                  "end": 190,
                  "start": 179,
                },
                "value": "empty_string",
              },
              "parent": [Circular],
              "type": "string",
              "value": Object {
                "code": "",
                "range": Object {
                  "end": 194,
                  "start": 195,
                },
              },
            },
            Object {
              "key": Object {
                "range": Object {
                  "end": 210,
                  "start": 205,
                },
                "value": "string",
              },
              "parent": [Circular],
              "type": "string",
              "value": Object {
                "code": "string",
                "range": Object {
                  "end": 220,
                  "start": 215,
                },
              },
            },
            Object {
              "key": Object {
                "range": Object {
                  "end": 232,
                  "start": 231,
                },
                "value": "\\\\\\"",
              },
              "parent": [Circular],
              "type": "string",
              "value": Object {
                "code": "\\\\\\"",
                "range": Object {
                  "end": 237,
                  "start": 236,
                },
              },
            },
            Object {
              "key": Object {
                "range": Object {
                  "end": 253,
                  "start": 248,
                },
                "value": "number",
              },
              "parent": [Circular],
              "type": "number",
              "value": Object {
                "code": "0",
                "range": Object {
                  "end": 257,
                  "start": 257,
                },
              },
            },
            Object {
              "key": Object {
                "range": Object {
                  "end": 270,
                  "start": 267,
                },
                "value": "null",
              },
              "parent": [Circular],
              "type": "null",
              "value": Object {
                "code": "null",
                "range": Object {
                  "end": 277,
                  "start": 274,
                },
              },
            },
            Object {
              "key": Object {
                "range": Object {
                  "end": 291,
                  "start": 287,
                },
                "value": "false",
              },
              "parent": [Circular],
              "type": "boolean",
              "value": Object {
                "code": "false",
                "range": Object {
                  "end": 298,
                  "start": 294,
                },
              },
            },
            Object {
              "key": Object {
                "range": Object {
                  "end": 311,
                  "start": 308,
                },
                "value": "true",
              },
              "parent": [Circular],
              "type": "boolean",
              "value": Object {
                "code": "true",
                "range": Object {
                  "end": 318,
                  "start": 315,
                },
              },
            },
          ],
          "type": "object",
          "value": Object {
            "code": "{
      \\"empty_string\\": \\"\\",
      \\"string\\": \\"string\\",
      \\"\\\\\\"\\":\\"\\\\\\"\\",
      \\"number\\": 0,
      \\"null\\": null,
      \\"false\\":false,
      \\"true\\": true
    }",
            "range": Object {
              "end": 324,
              "start": 170,
            },
          },
        },
        Object {
          "key": 1,
          "parent": [Circular],
          "properties": Array [
            Object {
              "key": 0,
              "parent": [Circular],
              "type": "number",
              "value": Object {
                "code": "1",
                "range": Object {
                  "end": 332,
                  "start": 332,
                },
              },
            },
            Object {
              "key": 1,
              "parent": [Circular],
              "type": "string",
              "value": Object {
                "code": "2",
                "range": Object {
                  "end": 335,
                  "start": 335,
                },
              },
            },
            Object {
              "key": 2,
              "parent": [Circular],
              "type": "null",
              "value": Object {
                "code": "null",
                "range": Object {
                  "end": 341,
                  "start": 338,
                },
              },
            },
            Object {
              "key": 3,
              "parent": [Circular],
              "type": "boolean",
              "value": Object {
                "code": "false",
                "range": Object {
                  "end": 347,
                  "start": 343,
                },
              },
            },
            Object {
              "key": 4,
              "parent": [Circular],
              "type": "boolean",
              "value": Object {
                "code": "true",
                "range": Object {
                  "end": 352,
                  "start": 349,
                },
              },
            },
          ],
          "type": "array",
          "value": Object {
            "code": "[1,\\"2\\",null,false,true]",
            "range": Object {
              "end": 353,
              "start": 331,
            },
          },
        },
        Object {
          "key": 2,
          "parent": [Circular],
          "type": "number",
          "value": Object {
            "code": "1",
            "range": Object {
              "end": 360,
              "start": 360,
            },
          },
        },
        Object {
          "key": 3,
          "parent": [Circular],
          "type": "string",
          "value": Object {
            "code": "2",
            "range": Object {
              "end": 363,
              "start": 363,
            },
          },
        },
        Object {
          "key": 4,
          "parent": [Circular],
          "type": "null",
          "value": Object {
            "code": "null",
            "range": Object {
              "end": 369,
              "start": 366,
            },
          },
        },
        Object {
          "key": 5,
          "parent": [Circular],
          "type": "boolean",
          "value": Object {
            "code": "false",
            "range": Object {
              "end": 375,
              "start": 371,
            },
          },
        },
        Object {
          "key": 6,
          "parent": [Circular],
          "type": "boolean",
          "value": Object {
            "code": "true",
            "range": Object {
              "end": 380,
              "start": 377,
            },
          },
        },
      ],
      "type": "array",
      "value": Object {
        "code": "[
    {
      \\"empty_string\\": \\"\\",
      \\"string\\": \\"string\\",
      \\"\\\\\\"\\":\\"\\\\\\"\\",
      \\"number\\": 0,
      \\"null\\": null,
      \\"false\\":false,
      \\"true\\": true
    },
    [1,\\"2\\",null,false,true],
    1,\\"2\\",null,false,true
  ]",
        "range": Object {
          "end": 384,
          "start": 164,
        },
      },
    },
  ],
  "type": "object",
  "value": Object {
    "code": "{
  \\"empty_string\\": \\"\\",
  \\"string\\": \\"string\\",
  \\"\\\\\\"\\":\\"\\\\\\"\\",
  \\"number\\": 0,
  \\"null\\": null,
  \\"false\\":false,
  \\"true\\": true,
  \\"array\\": [],
  \\"object\\": {},
  \\"deep\\": [
    {
      \\"empty_string\\": \\"\\",
      \\"string\\": \\"string\\",
      \\"\\\\\\"\\":\\"\\\\\\"\\",
      \\"number\\": 0,
      \\"null\\": null,
      \\"false\\":false,
      \\"true\\": true
    },
    [1,\\"2\\",null,false,true],
    1,\\"2\\",null,false,true
  ]
}",
    "range": Object {
      "end": 386,
      "start": 0,
    },
  },
}
`;

exports[`fixtures test array.json 1`] = `
Object {
  "key": null,
  "parent": null,
  "properties": Array [
    Object {
      "key": 0,
      "parent": [Circular],
      "type": "number",
      "value": Object {
        "code": "1",
        "range": Object {
          "end": 1,
          "start": 1,
        },
      },
    },
    Object {
      "key": 1,
      "parent": [Circular],
      "type": "string",
      "value": Object {
        "code": "2",
        "range": Object {
          "end": 4,
          "start": 4,
        },
      },
    },
    Object {
      "key": 2,
      "parent": [Circular],
      "type": "null",
      "value": Object {
        "code": "null",
        "range": Object {
          "end": 10,
          "start": 7,
        },
      },
    },
    Object {
      "key": 3,
      "parent": [Circular],
      "type": "boolean",
      "value": Object {
        "code": "false",
        "range": Object {
          "end": 16,
          "start": 12,
        },
      },
    },
    Object {
      "key": 4,
      "parent": [Circular],
      "type": "boolean",
      "value": Object {
        "code": "true",
        "range": Object {
          "end": 21,
          "start": 18,
        },
      },
    },
  ],
  "type": "array",
  "value": Object {
    "code": "[1,\\"2\\",null,false,true]",
    "range": Object {
      "end": 22,
      "start": 0,
    },
  },
}
`;

exports[`fixtures test boolean.json 1`] = `
Object {
  "key": null,
  "parent": null,
  "type": "boolean",
  "value": Object {
    "code": "false",
    "range": Object {
      "end": 4,
      "start": 0,
    },
  },
}
`;

exports[`fixtures test null.json 1`] = `
Object {
  "key": null,
  "parent": null,
  "type": "null",
  "value": Object {
    "code": "null",
    "range": Object {
      "end": 3,
      "start": 0,
    },
  },
}
`;

exports[`fixtures test number.json 1`] = `
Object {
  "key": null,
  "parent": null,
  "type": "number",
  "value": Object {
    "code": "123",
    "range": Object {
      "end": 2,
      "start": 0,
    },
  },
}
`;

exports[`fixtures test string.json 1`] = `
Object {
  "key": null,
  "parent": null,
  "type": "string",
  "value": Object {
    "code": "string",
    "range": Object {
      "end": 6,
      "start": 1,
    },
  },
}
`;
